Armadillo/src/main/java/com/scribd/armadillo/download/drm/ArmadilloDrmSessionManagerProvider.kt

```diff
@@ -0,0 +1,136 @@
+
+    private val lock = Any()
+
+    @GuardedBy("lock")
+    private var drmConfiguration: DrmConfiguration? = null
+
+    @GuardedBy("lock")
+    private var manager: DrmSessionManager? = null
+    private val drmHttpDataSourceFactory: DataSource.Factory? = null
+    private val userAgent: String? = null
+
+    // End DefaultDrmSessionManagerProvider fields //
+    /////////////////////////////////////////////////
+
+    companion object {
+        const val TAG = "ArmadilloDrmSMProvider"
+    }
+
+    /**  Identical to DefaultDrmSessionManagerProvider method, */
+    override fun get(mediaItem: MediaItem): DrmSessionManager {
+        Assertions.checkNotNull(mediaItem.localConfiguration)
+        val drmConfiguration = mediaItem.localConfiguration!!.drmConfiguration
+        if (drmConfiguration != null && Util.SDK_INT >= 18) {
+            synchronized(this.lock) {
+                if (!Util.areEqual(drmConfiguration, this.drmConfiguration)) {
+                    this.drmConfiguration = drmConfiguration
+                    this.manager = this.createManager(drmConfiguration)
+                }
+                return Assertions.checkNotNull<DrmSessionManager?>(this.manager)
+            }
+        } else {
+            return DrmSessionManager.DRM_UNSUPPORTED
+        }
+    }
+
+    /** Near identical to DefaultDrmSessionManagerProvider method, except for the indicated line */
+    private fun createManager(drmConfiguration: DrmConfiguration): DrmSessionManager {
+        val dataSourceFactory = this.drmHttpDataSourceFactory
+            ?: DefaultHttpDataSource.Factory().setUserAgent(this.userAgent)
+        val httpDrmCallback = HttpMediaDrmCallback(if (drmConfiguration.licenseUri == null) null else drmConfiguration.licenseUri.toString(), drmConfiguration.forceDefaultLicenseUri, (dataSourceFactory)!!)
+        val var4: UnmodifiableIterator<*> = drmConfiguration.licenseRequestHeaders.entries.iterator()
+
+        while (var4.hasNext()) {
+            val entry: Map.Entry<String, String> = var4.next() as Map.Entry<String, String>
+            httpDrmCallback.setKeyRequestProperty(entry.key, entry.value)
+        }
+
+        val drmSessionManager = DefaultDrmSessionManager.Builder()
+            //this line is different, changing the DrmProvider
+            .setUuidAndExoMediaDrmProvider(drmConfiguration.scheme, ArmadilloDrmProvider(stateStore))
+            .setMultiSession(drmConfiguration.multiSession)
+            .setPlayClearSamplesWithoutKeys(drmConfiguration.playClearContentWithoutKey)
+            .setUseDrmSessionsForClearContent(*Ints.toArray(drmConfiguration.forcedSessionTrackTypes))
+            .build(httpDrmCallback)
+        drmSessionManager.setMode(DefaultDrmSessionManager.MODE_PLAYBACK, drmConfiguration.keySetId)
+        return drmSessionManager
+    }
+
+    /** New original provider for this class */
+    class ArmadilloDrmProvider(private val stateStore: StateStore.Modifier) : ExoMediaDrm.Provider {
```

Comment:
Probably makes more sense for this to be a separate class in its own file than an inner class

Reply:
I felt that the large amount of single use classes  in separate files made everything very hard to follow.  It was very hard to figure out how to build this feature.

Reply:
>I felt that the large amount of single use classes in separate files made everything very hard to follow. It was very hard to figure out how to build this feature.

This x1000. This is what I meant by my audio player being a more simplified interpretation of armadillo and could be a good reference of what exactly is being done

Reply:
Yeah if i were to refactor this project, i would use some Facades, layers or something to help indicate hierarchy and order.